src/main/java/com/google/api/gax/bundling/ThresholdBundler.java

```diff
@@ -49,64 +48,83 @@
+  private class ReleaseResourcesFunction implements Function<Object, Void> {
+    private final E bundle;
+
+    private ReleaseResourcesFunction(E bundle) {
+      this.bundle = bundle;
+    }
+
+    @Override
+    public Void apply(Object input) {
+      flowController.release(bundle);
+      return null;
+    }
+  }
+
+  private final Runnable pushCurrentBundleRunnable =
+      new Runnable() {
+        @Override
+        public void run() {
+          pushCurrentBundle();
+        }
+      };
+
+  private final ArrayList<BundlingThreshold<E>> thresholds;
+  private final ScheduledExecutorService executor;
   private final Duration maxDelay;
+  private final ThresholdBundleReceiver<E> receiver;
   private final BundlingFlowController<E> flowController;
   private final BundleMerger<E> bundleMerger;
 
-  private final Lock lock = new ReentrantLock();
-  private final Condition bundleCondition = lock.newCondition();
-  private BundleState currentBundleState;
-  private Queue<E> closedBundles = new ArrayDeque<>();
-
-  private ThresholdBundler(
-      ImmutableList<BundlingThreshold<E>> thresholds,
-      Duration maxDelay,
-      BundlingFlowController<E> flowController,
-      BundleMerger<E> bundleMerger) {
-    this.thresholdPrototypes = copyResetThresholds(Preconditions.checkNotNull(thresholds));
-    this.maxDelay = maxDelay;
-    this.flowController = Preconditions.checkNotNull(flowController);
-    this.bundleMerger = Preconditions.checkNotNull(bundleMerger);
-    this.currentBundleState = null;
+  // Invariant:
+  // - lock gates all accesses to members below
+  // - currentOpenBundle and currentAlarmFuture are either both null or both non-null
+  private final ReentrantLock lock = new ReentrantLock();
+  private E currentOpenBundle;
+  private Future<?> currentAlarmFuture;
+
+  private ThresholdBundler(Builder<E> builder) {
+    this.thresholds = new ArrayList<>(builder.thresholds);
+    this.executor = Preconditions.checkNotNull(builder.executor);
+    this.maxDelay = Preconditions.checkNotNull(builder.maxDelay);
+    this.receiver = Preconditions.checkNotNull(builder.receiver);
+    this.flowController = Preconditions.checkNotNull(builder.flowController);
+    this.bundleMerger = Preconditions.checkNotNull(builder.bundleMerger);
+
+    resetThresholds();
   }
 
-  /** Builder for a ThresholdBundler. */
-  public static final class Builder<E> {
-    private List<BundlingThreshold<E>> thresholds;
+  public static class Builder<E> {
+    private Collection<BundlingThreshold<E>> thresholds;
+    private ScheduledExecutorService executor;
     private Duration maxDelay;
+    private ThresholdBundleReceiver<E> receiver;
     private BundlingFlowController<E> flowController;
     private BundleMerger<E> bundleMerger;
 
-    private Builder() {
-      thresholds = Lists.newArrayList();
+    private Builder() {}
+
+    public Builder<E> setThresholds(Collection<BundlingThreshold<E>> thresholds) {
```

Comment:
You're losing docs on the Builder here. Also, you might as well keep the functions in the same order as the original.